synapse/storage/databases/state/store.py

```diff
@@ -237,7 +257,142 @@ def _get_state_for_group_using_cache(
 
         return state_filter.filter_state(state_dict_ids), not missing_types
 
+    async def _get_state_groups_from_cache(
+        self, state_groups: Iterable[int], state_filter: StateFilter
+    ) -> Tuple[Dict[int, MutableStateMap[str]], Set[int]]:
+        """TODO
+
+        Returns:
+            A map from each state_group to the complete/incomplete state map (filled in by cached
+            values) and the set of incomplete groups
+        """
+        member_filter, non_member_filter = state_filter.get_member_split()
+
+        # Now we look them up in the member and non-member caches
+        (
+            non_member_state,
+            incomplete_groups_nm,
+        ) = self._get_state_for_groups_using_cache(
+            state_groups, self._state_group_cache, state_filter=non_member_filter
+        )
+
+        (member_state, incomplete_groups_m) = self._get_state_for_groups_using_cache(
+            state_groups, self._state_group_members_cache, state_filter=member_filter
+        )
+
+        state = dict(non_member_state)
+        for state_group in state_groups:
+            state[state_group].update(member_state[state_group])
+
+        # We may have only got one of the events for the group
+        incomplete_groups = incomplete_groups_m | incomplete_groups_nm
+
+        return (state, incomplete_groups)
+
+    @cancellable
+    @trace
+    @tag_args
+    async def _get_state_for_client_filtering(
+        self, groups: Iterable[int], user_id_viewing_events: str
+    ) -> Dict[int, StateMap[str]]:
+        """
+        TODO
+        """
+
+        def _get_state_for_client_filtering_txn(
+            txn: LoggingTransaction, groups: Iterable[int]
+        ) -> Mapping[int, StateMap[str]]:
+            sql = """
+                WITH RECURSIVE sgs(state_group) AS (
+                    VALUES(?::bigint)
+                    UNION ALL
+                    SELECT prev_state_group FROM state_group_edges e, sgs s
+                    WHERE s.state_group = e.state_group
+                )
+                SELECT
+                    type, state_key, event_id
+                FROM state_groups_state
+                WHERE
+                    state_group IN (
+                        SELECT state_group FROM sgs
+                    )
+                    AND (type = ? AND state_key = ?)
+                ORDER BY
+                    type,
+                    state_key,
+                    -- Use the lastest state in the chain (highest numbered state_group in the chain)
+                    state_group DESC
+                LIMIT 1
```

Comment:
### Before

One query that takes ~80ms

```
EXPLAIN ANALYZE WITH RECURSIVE sgs(state_group) AS (
    VALUES(739988088::bigint)
    UNION ALL
    SELECT prev_state_group FROM state_group_edges e, sgs s
    WHERE s.state_group = e.state_group
)
SELECT
    state_group, type, state_key, event_id
FROM state_groups_state
WHERE
    state_group IN (
        SELECT state_group FROM sgs
    )
    AND ((type = 'm.room.history_visibility' AND state_key = '') OR (type = 'm.room.member' AND state_key = '@madlittlemods:matrix.org'))
ORDER BY type, state_key, state_group DESC;
                                                                                       QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=5274.51..5274.51 rows=1 width=87) (actual time=70.977..70.980 rows=2 loops=1)
   Sort Key: state_groups_state.type, state_groups_state.state_key, state_groups_state.state_group DESC
   Sort Method: quicksort  Memory: 25kB
   CTE sgs
     ->  Recursive Union  (cost=0.00..284.28 rows=101 width=8) (actual time=0.003..0.475 rows=59 loops=1)
           ->  Result  (cost=0.00..0.01 rows=1 width=8) (actual time=0.002..0.002 rows=1 loops=1)
           ->  Nested Loop  (cost=0.57..28.23 rows=10 width=8) (actual time=0.007..0.007 rows=1 loops=59)
                 ->  WorkTable Scan on sgs s  (cost=0.00..0.20 rows=10 width=8) (actual time=0.000..0.000 rows=1 loops=59)
                 ->  Index Only Scan using state_group_edges_unique_idx on state_group_edges e  (cost=0.57..2.79 rows=1 width=16) (actual time=0.006..0.007 rows=1 loops=59)
                       Index Cond: (state_group = s.state_group)
                       Heap Fetches: 58
   ->  Nested Loop  (cost=3.11..4990.22 rows=1 width=87) (actual time=0.776..70.965 rows=2 loops=1)
         ->  HashAggregate  (cost=2.27..3.28 rows=101 width=8) (actual time=0.516..0.528 rows=59 loops=1)
               Group Key: sgs.state_group
               Batches: 1  Memory Usage: 24kB
               ->  CTE Scan on sgs  (cost=0.00..2.02 rows=101 width=8) (actual time=0.004..0.491 rows=59 loops=1)
         ->  Index Scan using state_groups_state_type_idx on state_groups_state  (cost=0.84..49.37 rows=1 width=87) (actual time=0.009..1.194 rows=0 loops=59)
               Index Cond: (state_group = sgs.state_group)
               Filter: (((type = 'm.room.history_visibility'::text) AND (state_key = ''::text)) OR ((type = 'm.room.member'::text) AND (state_key = '@madlittlemods:matrix.org'::text)))
               Rows Removed by Filter: 1495
 Planning Time: 2.217 ms
 Execution Time: 71.040 ms
(22 rows)

Time: 73.891 ms
```


### After

Two query that each take about ~4ms

```
EXPLAIN ANALYZE WITH RECURSIVE sgs(state_group) AS (
    VALUES(739988088::bigint)
    UNION ALL
    SELECT prev_state_group FROM state_group_edges e, sgs s
    WHERE s.state_group = e.state_group
)
SELECT
    state_group, type, state_key, event_id
FROM state_groups_state
WHERE
    state_group IN (
        SELECT state_group FROM sgs
    )
    AND (type = 'm.room.member' AND state_key = '@madlittlemods:matrix.org')
ORDER BY
    type,
    state_key,
    -- Use the lastest state in the chain
    state_group DESC
LIMIT 1;
                                                                                 QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=779.70..779.70 rows=1 width=87) (actual time=1.248..1.251 rows=1 loops=1)
   CTE sgs
     ->  Recursive Union  (cost=0.00..284.28 rows=101 width=8) (actual time=0.003..0.653 rows=59 loops=1)
           ->  Result  (cost=0.00..0.01 rows=1 width=8) (actual time=0.001..0.002 rows=1 loops=1)
           ->  Nested Loop  (cost=0.57..28.23 rows=10 width=8) (actual time=0.009..0.010 rows=1 loops=59)
                 ->  WorkTable Scan on sgs s  (cost=0.00..0.20 rows=10 width=8) (actual time=0.000..0.000 rows=1 loops=59)
                 ->  Index Only Scan using state_group_edges_unique_idx on state_group_edges e  (cost=0.57..2.79 rows=1 width=16) (actual time=0.008..0.009 rows=1 loops=59)
                       Index Cond: (state_group = s.state_group)
                       Heap Fetches: 58
   ->  Sort  (cost=495.42..495.42 rows=1 width=87) (actual time=1.246..1.247 rows=1 loops=1)
         Sort Key: state_groups_state.state_group DESC
         Sort Method: quicksort  Memory: 25kB
         ->  Nested Loop  (cost=3.11..495.41 rows=1 width=87) (actual time=0.925..1.239 rows=1 loops=1)
               ->  HashAggregate  (cost=2.27..3.28 rows=101 width=8) (actual time=0.717..0.729 rows=59 loops=1)
                     Group Key: sgs.state_group
                     Batches: 1  Memory Usage: 24kB
                     ->  CTE Scan on sgs  (cost=0.00..2.02 rows=101 width=8) (actual time=0.005..0.680 rows=59 loops=1)
               ->  Index Scan using state_groups_state_type_idx on state_groups_state  (cost=0.84..4.86 rows=1 width=87) (actual time=0.008..0.008 rows=0 loops=59)
                     Index Cond: ((state_group = sgs.state_group) AND (type = 'm.room.member'::text) AND (state_key = '@madlittlemods:matrix.org'::text))
 Planning Time: 2.611 ms
 Execution Time: 1.326 ms
(21 rows)

Time: 4.614 ms
```

```
WITH RECURSIVE sgs(state_group) AS (
    VALUES(739988088::bigint)
    UNION ALL
    SELECT prev_state_group FROM state_group_edges e, sgs s
    WHERE s.state_group = e.state_group
)
SELECT
    state_group, type, state_key, event_id
FROM state_groups_state
WHERE
    state_group IN (
        SELECT state_group FROM sgs
    )
    AND (type = 'm.room.history_visibility' AND state_key = '')
ORDER BY
    type,
    state_key,
    -- Use the lastest state in the chain
    state_group DESC
LIMIT 1;

                                                                                 QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=779.70..779.70 rows=1 width=87) (actual time=0.961..0.964 rows=1 loops=1)
   CTE sgs
     ->  Recursive Union  (cost=0.00..284.28 rows=101 width=8) (actual time=0.003..0.458 rows=59 loops=1)
           ->  Result  (cost=0.00..0.01 rows=1 width=8) (actual time=0.002..0.002 rows=1 loops=1)
           ->  Nested Loop  (cost=0.57..28.23 rows=10 width=8) (actual time=0.007..0.007 rows=1 loops=59)
                 ->  WorkTable Scan on sgs s  (cost=0.00..0.20 rows=10 width=8) (actual time=0.000..0.000 rows=1 loops=59)
                 ->  Index Only Scan using state_group_edges_unique_idx on state_group_edges e  (cost=0.57..2.79 rows=1 width=16) (actual time=0.006..0.006 rows=1 loops=59)
                       Index Cond: (state_group = s.state_group)
                       Heap Fetches: 58
   ->  Sort  (cost=495.42..495.42 rows=1 width=87) (actual time=0.960..0.961 rows=1 loops=1)
         Sort Key: state_groups_state.state_group DESC
         Sort Method: quicksort  Memory: 25kB
         ->  Nested Loop  (cost=3.11..495.41 rows=1 width=87) (actual time=0.675..0.953 rows=1 loops=1)
               ->  HashAggregate  (cost=2.27..3.28 rows=101 width=8) (actual time=0.503..0.514 rows=59 loops=1)
                     Group Key: sgs.state_group
                     Batches: 1  Memory Usage: 24kB
                     ->  CTE Scan on sgs  (cost=0.00..2.02 rows=101 width=8) (actual time=0.004..0.476 rows=59 loops=1)
               ->  Index Scan using state_groups_state_type_idx on state_groups_state  (cost=0.84..4.86 rows=1 width=87) (actual time=0.007..0.007 rows=0 loops=59)
                     Index Cond: ((state_group = sgs.state_group) AND (type = 'm.room.history_visibility'::text) AND (state_key = ''::text))
 Planning Time: 1.648 ms
 Execution Time: 1.018 ms
(21 rows)

Time: 3.280 ms
```

Reply:
I don't have exact details to be confident in why this is faster but my thinking was we will be trying to do less by matching quicker because of one criteria at a time and `LIMIT 1` where it can bail early after one successful `state_groups_state` match. And not doing any `DISTINCT` weirdness that probably keeps track of values.

Plus the query planner shows us that we're not throwing away 88k of rows filtering them away anymore.

Reply:
Oooh, that is very interesting. I think what is going on here is that the `OR` condition means that the query planner is no longer just pulling out the specific type/state_key pairs it needs, its instead pulling out the entire state group and filtering in memory. This has significant impact for rooms with large state groups, obviously.

One alternate proposal is instead of using `OR` conditions is to instead do `UNION`, e.g.:

```sql
EXPLAIN ANALYZE WITH RECURSIVE sgs(state_group) AS (
    VALUES(739988088::bigint)
    UNION ALL
    SELECT prev_state_group FROM state_group_edges e, sgs s
    WHERE s.state_group = e.state_group
)
(
    SELECT DISTINCT ON (type, state_key)
        state_group, type, state_key, event_id
    FROM state_groups_state
    INNER JOIN sgs USING (state_group)
    WHERE (type = 'm.room.member' AND state_key = '@madlittlemods:matrix.org')
    ORDER BY type, state_key, state_group DESC
) UNION (
    SELECT DISTINCT ON (type, state_key)
        state_group, type, state_key, event_id
    FROM state_groups_state
    INNER JOIN sgs USING (state_group)
    WHERE (type = 'm.room.history_visibility' AND state_key = '')
    ORDER BY type, state_key, state_group DESC
);
```

This has the advantage of being a single query, and reusing the work for pulling out the state edges.

Reply:
I think we can actually `_get_state_groups_from_groups_txn` implementation to do the above if we get a state filter? Would likely be a much smaller change?

Reply:
@erikjohnston Have you tested your query? It does seem like an improvement over `develop` but takes longer and about ~20ms after Postgres cache is warm still.

https://explain.depesz.com/s/cyaR

<details>
<summary>Raw query plan</summary>

```
                                                                                      QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 HashAggregate  (cost=1246.10..1246.61 rows=51 width=104) (actual time=28.398..28.403 rows=0 loops=1)
   Group Key: state_groups_state.state_group, state_groups_state.type, state_groups_state.state_key, state_groups_state.event_id
   Batches: 1  Memory Usage: 24kB
   CTE sgs
     ->  Recursive Union  (cost=0.00..515.03 rows=101 width=8) (actual time=0.003..1.594 rows=1 loops=1)
           ->  Result  (cost=0.00..0.01 rows=1 width=8) (actual time=0.001..0.002 rows=1 loops=1)
           ->  Nested Loop  (cost=0.28..51.30 rows=10 width=8) (actual time=1.590..1.590 rows=0 loops=1)
                 ->  WorkTable Scan on sgs s  (cost=0.00..0.20 rows=10 width=8) (actual time=0.001..0.002 rows=1 loops=1)
                 ->  Index Only Scan using state_group_edges_unique_idx on state_group_edges e  (cost=0.28..5.10 rows=1 width=16) (actual time=1.585..1.585 rows=0 loops=1)
                       Index Cond: (state_group = s.state_group)
                       Heap Fetches: 0
   ->  Append  (cost=0.28..730.56 rows=51 width=104) (actual time=28.392..28.395 rows=0 loops=1)
         ->  Unique  (cost=0.28..304.38 rows=1 width=95) (actual time=18.021..18.021 rows=0 loops=1)
               ->  Nested Loop  (cost=0.28..304.38 rows=1 width=95) (actual time=18.020..18.020 rows=0 loops=1)
                     Join Filter: (state_groups_state.state_group = sgs.state_group)
                     ->  Index Scan Backward using state_groups_state_type_idx on state_groups_state  (cost=0.28..301.10 rows=1 width=95) (actual time=18.019..18.019 rows=0 loops=1)
                           Index Cond: ((type = 'm.room.member'::text) AND (state_key = '@madlittlemods:matrix.org'::text))
                     ->  CTE Scan on sgs  (cost=0.00..2.02 rows=101 width=8) (never executed)
         ->  Unique  (cost=425.28..425.41 rows=50 width=95) (actual time=10.369..10.371 rows=0 loops=1)
               ->  Sort  (cost=425.28..425.41 rows=50 width=95) (actual time=10.368..10.370 rows=0 loops=1)
                     Sort Key: state_groups_state_1.state_group DESC
                     Sort Method: quicksort  Memory: 25kB
                     ->  Hash Join  (cost=300.39..423.87 rows=50 width=95) (actual time=9.244..9.245 rows=0 loops=1)
                           Hash Cond: (state_groups_state_1.state_group = sgs_1.state_group)
                           ->  Bitmap Heap Scan on state_groups_state state_groups_state_1  (cost=297.11..419.88 rows=56 width=95) (actual time=0.921..5.960 rows=232 loops=1)
                                 Recheck Cond: ((type = 'm.room.history_visibility'::text) AND (state_key = ''::text))
                                 Heap Blocks: exact=62
                                 ->  Bitmap Index Scan on state_groups_state_type_idx  (cost=0.00..297.09 rows=56 width=0) (actual time=0.904..0.905 rows=232 loops=1)
                                       Index Cond: ((type = 'm.room.history_visibility'::text) AND (state_key = ''::text))
                           ->  Hash  (cost=2.02..2.02 rows=101 width=8) (actual time=2.140..2.141 rows=1 loops=1)
                                 Buckets: 1024  Batches: 1  Memory Usage: 9kB
                                 ->  CTE Scan on sgs sgs_1  (cost=0.00..2.02 rows=101 width=8) (actual time=0.004..1.595 rows=1 loops=1)
 Planning Time: 25.952 ms
 Execution Time: 33.217 ms
(34 rows)

Time: 141.669 ms
```

</details>

Reply:
I takes only a few ms for me :/

Reply:
Done in v2 -> https://github.com/matrix-org/synapse/pull/14527